quickwit-serve/src/rest.rs

```diff
@@ -0,0 +1,296 @@
+    num_microsecs: u64,
+}
+
+impl From<SearchResult> for SearchResultJson {
+    fn from(search_result: SearchResult) -> Self {
+        let hits: Vec<serde_json::Value> = search_result.hits.into_iter()
+            .map(|hit| {
+                match serde_json::from_str(&hit.json) {
+                    Ok(hit_json) => hit_json,
+                    Err(invalid_json) => {
+                        error!(err=?invalid_json, "Invalid json in hit");
+                        serde_json::json!({})
+                    }
+                }
+            })
+            .collect();
+        SearchResultJson {
+            num_hits: search_result.num_hits,
+            hits,
+            num_microsecs: search_result.elapsed_time_micros,
+        }
+    }
+}
+
+async fn search_endpoint<TSearchService: SearchService>(
+    index_id: String,
+    search_request: SearchRequestQueryString,
+    search_service: &TSearchService,
+) -> Result<SearchResultJson, ApiError> {
+    let search_request = grpc::SearchRequest {
+        index_id,
+        query: search_request.query,
+        start_timestamp: search_request.start_timestamp,
+        end_timestamp: search_request.end_timestamp,
+        max_hits: search_request.max_hits,
+        start_offset: search_request.start_offset,
+    };
+    let search_result = search_service.root_search(search_request).await?;
+    let search_result_json = SearchResultJson::from(search_result);
+    Ok(search_result_json)
+}
+
+fn search_filter(
+) -> impl Filter<Extract = (String, SearchRequestQueryString), Error = Rejection> + Clone {
+    warp::path!("api" / "v1" / String / "search")
+        .and(warp::get())
+        .and(serde_qs::warp::query(serde_qs::Config::default()))
+}
+
+async fn search<TSearchService: SearchService>(
+    index_id: String,
+    request: SearchRequestQueryString,
+    search_service: Arc<TSearchService>,
+) -> Result<impl warp::Reply, Infallible> {
+    info!(index_id=%index_id,request=?request, "search");
+    Ok(make_reply(
+        search_endpoint(index_id, request, &*search_service).await,
+    ))
+}
+
+/// Rest search handler.
+///
+/// It parses the search request from the
+pub fn search_handler<TSearchService: SearchService>(
+    search_service: Arc<TSearchService>,
+) -> impl Filter<Extract = impl warp::Reply, Error = Rejection> + Clone {
+    search_filter()
+        .and(warp::any().map(move || search_service.clone()))
+        .and_then(search)
+        .recover(recover_fn)
+}
+
+fn make_reply<T: serde::Serialize>(result: Result<T, ApiError>) -> reply::WithStatus<reply::Json> {
+    match result {
+        Ok(success) => reply::with_status(reply::json(&success), StatusCode::OK),
+        Err(error) => reply::with_status(reply::json(&error), error.http_status_code()),
+    }
+}
+
+/// This function returns a formated error based on the given rejection reason.
+async fn recover_fn(rejection: Rejection) -> Result<impl Reply, Rejection> {
+    match rejection.find::<serde_qs::Error>() {
```

Comment:
Rejection can have other causes that serde_qs error, I would add a todo here to handle all cases. Cf [example](https://github.com/seanmonstar/warp/blob/master/examples/rejections.rs#L84-L90)

I tried to make a "HEAD" request in the test and I receive a 404 response but normally I should have had a 405 method not allowed.